Add Unsafe Rust Deep Dive (#2806)
google/comprehensive-rust: src/running-the-course/course-structure.md

{{%course outline Concurrency}}

> ### Unsafe (Work in Progress)

Reviewer: WDYT about "Unsafe Rust Fundamentals"?

> ### Unsafe (Work in Progress)

Reviewer: To help pick a better name, let me explain the content and the goals for this deep dive.

Internally at Google we have introduced a mandatory code review process for all unsafe Rust code. This is proving to be too onerous for engineers: even if code reviews complete quickly, people feel like we took away some of their agency.

Solution:

* ensure that engineers understand foundations of unsafe code, "the rules of the game" so to speak:

  - what the dangers of unsafe Rust are,
  - where to find information about specific unsafe operations,
  - how to read unsafe contracts,
  - how to satisfy unsafe contracts,
  - how to document all of the above in code comments for future readers,

* ensure engineers can distinguish simple cases (e.g., cookie-cutter FFI) from complex cases (e.g., something like the zerocopy crate),

* teach enough specifics about basic unsafe operations and do enough exercises that upon completion of the deep dive we can trust students to self-certify correctness of the simple unsafe code, and pull in experienced unsafe reviewers for complex cases.

> ### Unsafe (Work in Progress)

Reviewer 2: I was doing some of those reviews while I was there, and indeed it was a sticking point! That plan sounds fantastic, and will be a benefit to the Rust community as a whole.

My takeaway from that explanation is that the course is about `unsafe`, as that appears or is implied in all three bullet points. Safety and soundness are implied, but secondary. I'd consider "Working with Unsafe" but it seems long. So perhaps "Unsafe" is OK after all.

> ### Unsafe (Work in Progress)

Author: Really encouraging to hear that you're supportive @djmitche. 

A pithy title seems like a fun thing to add, but I was wary about introducing too much into an initial review. That said, What about "Safe Unsafe" or "Unsafe Safely"?

> ### Unsafe (Work in Progress)

Reviewer 2: I like both of those, and they are nice and short! If I had to pick I'd prefer the first ("Safe Unsafe").

> ### Unsafe (Work in Progress)

Reviewer: > Safe Unsafe

I'd be a pedant and say that this misuses the word "safe", the correct term would be "Sound Unsafe"

> ### Unsafe (Work in Progress)

Reviewer: While "Safe and Sound" is cute, it limits discoverability in a culturally-independent way: if someone is looking for information about unsafe, why would they click on "Safe and Sound"?

> ### Unsafe (Work in Progress)

Author: > Sound Unsafe

I wonder if this verges on "Sounds Unsafe", which has a significantly more negative connotation.

The [Unsafe](../unsafe-deep-dive/welcome.md) deep dive is a two day class on
using the `unsafe` keyword correctly. It covers the fundamentals of what Rust's
safety guarantees are and why `unsafe` is needed, reviewing `unsafe` code, using

> FFI and building data structures.

Author: Perhaps we should have the data structures item on its own line?

{{%course outline Idiomatic Rust}}

### Unsafe (Work in Progress)

The [Unsafe](../unsafe-deep-dive/welcome.md) deep dive is a two-day class on the

> _unsafe_ Rust language. It covers the fundamentals of what Rust's safety
> guarantees are and why `unsafe` is needed, reviewing `unsafe` code, using FFI,
> and building data structures that the borrow checker would normally reject.

Reviewer: ```suggestion
_unsafe_ Rust language. It covers the fundamentals of Rust's safety
guarantees, the motivation for `unsafe`, review process for `unsafe` code, FFI basics,
and building data structures that the borrow checker would normally reject.
```

An attempt to make list items parallel each other better.
